File: weather.rb
#!/usr/bin/env ruby

require 'rubygems'
gem 'nokogiri'
require 'nokogiri'
require 'open-uri'
require 'REXML/document'

module Weather

  def salt
    doc = REXML::Document.new(open("http://sgs.salt/xml/salt-tcs-icd.xml").read)
    doubles = doc.elements.to_a("//DBL")
    strings = doc.elements.to_a("//String")
    dbl = Hash.new
    str = Hash.new
    doubles.each { |d| 
      key = d.elements["Name"].text
      val = d.elements["Val"].text
      dbl[key] = val.to_f
    }
    strings.each { |s|
      key = s.elements["Name"].text
      val = s.elements["Val"].text
      str[key] = val
    }
    ntemps = 7
    doc.elements.each("*/Cluster/Array/Dimsize") { |e| ntemps = e.text.to_i }
    temp = 0.0
    doc.elements.each("*/Cluster/Array/DBL/Val") { |e|
      temp = temp + e.text.to_f
    }
    temp = temp/ntemps
    
    wx = Hash.new
    wx["SAST"] = str["SAST"].split(' ')[1]
    wx["Date"] = str["SAST"].split(' ')[0]
    wx["Air Pressure"] = "%.1f" % (dbl["Air pressure"]*10.0)
    wx["Dewpoint"] = "%.2f" % dbl["Dewpoint"]
    wx["RH"] = "%.1f" % dbl["Rel Humidity"]
    wx["Wind Speed (30m)"] = "%.1f" % (dbl["Wind mag 30m"]*3.6)
    wx["Wind Speed (10m)"] = "%.1f" % (dbl["Wind mag 10m"]*3.6)
    wx["Wind Dir (30m)"] = "%.1f" % dbl["Wind dir 30m"]
    wx["Wind Dir (10m)"] = "%.1f" % dbl["Wind dir 10m"]
    wx["Temp"] = "%.1f" % temp
    t = temp - dbl["Dewpoint"]
    wx["T - DP"] = "%.1f" % t
    return wx
  end
  
  def salt_txt
    wx = IO.popen("curl -s http://www.salt.ac.za/~saltmet/weather.txt").readlines

    keys = wx[0].chomp.split("\t")
    latest = wx[-1].chomp.split("\t")
    recent = wx[-10].chomp.split("\t")
    first = wx[1].chomp.split("\t")
    
    now = Hash.new
    hour = Hash.new
    tenmin = Hash.new

    i = 0
    keys.each { |k|
      now[k] = latest[i]
      hour[k] = first[i]
      tenmin[k] = recent[i]
      i = i + 1
    }

    now["Delta DP"] = now["Dew Point"].to_f - hour["Dew Point"].to_f
    now["T - DP"] = now["Temp (2m)"].to_f - now["Dew Point"].to_f

    return now
  end

  def wasp
    doc = Nokogiri::HTML(open("http://swaspgateway.suth/"))
    wx = Hash.new
    wx["Sky"] = doc.xpath("//table")[0].xpath("//td")[10].content
    wx["T - DP"] = doc.xpath("//table")[0].xpath("//td")[9].content.to_f
    wx["RH"] = doc.xpath("//table")[0].xpath("//td")[8].content.to_f
    wx["Temp"] = doc.xpath("//table")[0].xpath("//td")[7].content.to_f
    wx["Wind dir"] = doc.xpath("//table")[0].xpath("//td")[6].content
    wx["Wind speed"] = doc.xpath("//table")[0].xpath("//td")[5].content.to_f
    wx["Rain"] = doc.xpath("//table")[0].xpath("//td")[4].content
    wx["UT"] = doc.xpath("//table")[0].xpath("//td")[3].content.chomp
    return wx
  end

  def ness
    doc = Nokogiri.HTML(open("http://ystar1.suth/weather_log.php"))
    wx = Hash.new
    wx["SAST"] = doc.xpath("//table")[13].xpath("//td")[53].content
    wx["RH"] = doc.xpath("//table")[13].xpath("//td")[54].content.to_f
    wx["Temp"] = doc.xpath("//table")[13].xpath("//td")[58].content.to_f
    wx["Wind speed"] = doc.xpath("//table")[13].xpath("//td")[59].content.to_f*3.6
    wx["Wind dir"] = doc.xpath("//table")[13].xpath("//td")[60].content.to_f
    return wx
  end

  def grav
    kan11 = Nokogiri.HTML(open("http://sg1.suth/tmp/kan11.htm"))
    kan16 = Nokogiri.HTML(open("http://sg1.suth/tmp/kan16.htm"))
    wx = Hash.new
    wx["SAST"] = kan11.xpath("//td")[12].content.split(' ')[1]
    wx["Date"] = kan11.xpath("//td")[12].content.split(' ')[0]
    wx["Temp"] = kan11.xpath("//td")[14].content.to_f
    wx["RH"] = kan11.xpath("//td")[15].content.to_f
    wx["Wind dir"] = kan16.xpath("//td")[13].content.to_i
    wx["Wind speed"] = kan16.xpath("//td")[14].content.to_f*3.6
    return wx
  end

end

if $0 == __FILE__
  include Weather
  wx = eval ARGV[0]
  wx.keys.sort.each { |k|
    puts "%20s %25s" % [k, wx[k]]
  }
end

File: weather.py
#!/usr/bin/env python

import xml
import sys
import html5lib
import urllib2
import xml.dom.minidom
from numpy import median, array
from xml_icd import *

from html5lib import treebuilders


def salt():
    wx = {}
    try:
        tcs = parseICD("http://sgs.salt/xml/salt-tcs-icd.xml")
        time = tcs['tcs xml time info']
        bms = tcs['bms external conditions']
        temps = array(bms['Temperatures'])

        # get temps and take median of the BMS output of 7 values
        wx["Temp"] = median(temps)

        # get time
        wx["SAST"] = time["SAST"].split()[1]
        wx["Date"] = time["SAST"].split()[0]

        # set up other values of interest
        wx["Air Pressure"] = bms["Air pressure"] * 10.0
        wx["Dewpoint"] = bms["Dewpoint"]
        wx["RH"] = bms["Rel Humidity"]
        wx["Wind Speed (30m)"] = bms["Wind mag 30m"] * 3.6
        wx["Wind Speed"] = bms["Wind mag 10m"] * 3.6
        wx["Wind Dir (30m)"] = bms["Wind dir 30m"]
        wx["Wind Dir"] = bms["Wind dir 10m"]
        wx["T - DP"] = wx["Temp"] - bms["Dewpoint"]
        wx["Raining"] = bms["Rain detected"]
        return wx
    except:
        return False


def wasp():
    wx = {}
    try:
        p = html5lib.HTMLParser(tree=treebuilders.getTreeBuilder("dom"))
        doc = p.parse(urllib2.urlopen("http://swaspgateway.suth/",
                                      timeout=1).read())
        t = doc.getElementsByTagName("table")[0]
        tds = t.getElementsByTagName("td")
        wx["Temp"] = float(tds[7].firstChild.nodeValue)
        if tds[10].firstChild.nodeValue == "RAIN":
            wx["Sky"] = "Rain"
            wx["Sky Temp"] = wx["Temp"]
        else:
            sky, stemp = tds[10].firstChild.nodeValue.split('(')
            stemp = stemp[0:-1]
            wx["Sky"] = sky
            wx["Sky Temp"] = stemp
        wx["T - DP"] = float(tds[9].firstChild.nodeValue)
        wx["RH"] = float(tds[8].firstChild.nodeValue)
        tds[6].normalize()
        wx["Wind Dir"] = tds[6].firstChild.nodeValue[1:]
        wx["Wind Speed"] = float(tds[5].firstChild.nodeValue)
        rain = tds[4].firstChild.nodeValue
        if rain == "DRY":
            wx["Raining"] = False
        else:
            wx["Raining"] = True
        wx["UT"] = tds[3].firstChild.nodeValue.strip()
        tds[31].normalize()
        wx["Status"] = tds[31].firstChild.nodeValue.strip()
        return wx
    except:
        return False


def grav():
    wx = {}
    p = html5lib.HTMLParser(tree=treebuilders.getTreeBuilder("dom"))
    kan11 = p.parse(urllib2.urlopen("http://sg1.suth/tmp/kan11.htm",
                                    timeout=1).read())
    kan16 = p.parse(urllib2.urlopen("http://sg1.suth/tmp/kan16.htm",
                                    timeout=1).read())
    kan11_tds = kan11.getElementsByTagName("td")
    kan16_tds = kan16.getElementsByTagName("td")
    wx["Date"], wx["UT"] = kan11_tds[12].firstChild.nodeValue.split()
    kan11_tds[14].normalize()
    kan11_tds[15].normalize()
    wx["Temp"] = float(kan11_tds[14].firstChild.nodeValue)
    wx["RH"] = float(kan11_tds[15].firstChild.nodeValue)
    kan16_tds[13].normalize()
    kan16_tds[14].normalize()
    wx["Wind Dir"] = int(kan16_tds[13].firstChild.nodeValue)
    wx["Wind Speed"] = float(kan16_tds[14].firstChild.nodeValue) * 3.6
    return wx


if __name__ == '__main__':
    if len(sys.argv) == 1:
        print "Usage: weather.py <salt|wasp|grav>"
    else:
        wx = eval("%s()" % sys.argv[1].lower())
        if wx:
            for k, v in sorted(wx.items()):
                print "%20s : \t %s" % (k, v)
        else:
            print "No information received."
